designer/client/designer.tsx

```diff
@@ -78,6 +80,18 @@ export default class Designer extends Component<Props, State> {
     const id = this.props.match?.params?.id;
     this.setState({ id });
     this.designerApi.fetchData(id).then((data) => {
+      const newFormItemNames = Object.getOwnPropertyNames(newFormJson);
+      newFormItemNames?.map((item) => {
+        if (!data?.hasOwnProperty(item)) {
+          data[item] = newFormJson[item];
+        }
+      });
+      const dataFormItems = Object.getOwnPropertyNames(data);
+      dataFormItems?.map((item) => {
+        if (!newFormJson?.hasOwnProperty(item)) {
+          delete data[item];
+        }
+      });
```

Comment:
seems like you're replacing the old values with the new values, then deleting entires that do not exist.

You can do with joi on the server. `schema.validate(response).options({ stripUnknown: true });` - So, if the joi schema does not recognise any keys, the resulting value will have the unknown keys removed. 

aside from that, this code has "side effects". The map methods are changing the `data` variable, it can make it hard to follow what's happening with the code when it's not "pure". `.map` should be used for creating a new array. I think you probably meant to use a for loop or `array.forEach`? I would do something like this

```suggestion
    const acceptedKeys = Object.keys(data); // 1
    const sanitisedNewFormJsonEntries = Object.entries(newFormJson).filter(([key, value]) => return acceptedKeys.include(key)) // 2
    const santitisedNewFormJson = Object.fromEntries(sanitisedNewFormJsonEntries) // 3
    const newData = {
    ...data,
    ...sanitisedNewFormJson
    } // 4
```
1. Get just the keys (as an array) of data
2. get the entries of the new data. an entry is [key, value], so entries returns an array of these tuples. filter the entries - if it exists on acceptedKeys, keep it. `sanitisedNewFormJsonEntries` should only include key/value pairs that exist on the original data. 
3. `Object.fromEntries` turns the array of key/value pairs back into an object
4. destructure `data` as well as `sanitisedNewFormJson` (must be in that order, so that any key/values that have been declared before will get overwritten)